### DISCUSSION TITLE
How should I provide additional params to generators?

### DISCUSSION THREAD

**Discussion by Synchro** [spatie/laravel-medialibrary#2152] [Q&A] [answered]
I'm looking at creating custom generators (PR incoming!), and looking for a way to provide some customisation of the image generation process. For example, how would I provide an extra parameter to specify that images should be generated in black & white, or that video files should make a composite thumbnail made from multiple frames?

The only way I can see so far is to provide global settings via `.env`, which seems a bit blunt.

Note that I'm not asking how to implement those things (that's my problem!), just how to pass in the parameters to control them.

**Comment by Synchro** [comment]
Related question, I'm thinking about having alternative implementations of thumbnail generators – is it reasonable to have more than one generator for a given file type? If so, how does ML choose between them? Via a developer-chosen ordering?

**Reply by freekmurze** [reply]
I'm thinking the first class that says it can convert the image, will convert the image.

**Reply by Synchro** [reply]
That's what I gathered, so I think it would be easiest to combine alternative renderers into a single extension and use config to select the one you want.

Regarding the first question, I implemented it using config for the PR I just submitted.

**Comment by Synchro** [comment, marked as answer]
This was addressed in #2156
